base/shell.jl

```diff
@@ -253,3 +253,84 @@ julia> Base.shell_escape_posixly("echo", "this", "&&", "that")
 """
 shell_escape_posixly(args::AbstractString...) =
     sprint(print_shell_escaped_posixly, args...)
+
+
+function print_shell_escaped_wincmd(io::IO, args::AbstractString...)
```

Comment:
Where does your `print_shell_escaped_CMDly` take care of the [quote flag that CMD uses when parsing its input](https://stackoverflow.com/a/4095133/1990689)? This doesn't look right and needs more careful testing with stings that contain all possible permutations of " and ^ and space (after both even and odd numbers of ").

May I also suggest that the shell_escaped_winsomely function returns an array of individually escaped argument strings, and does not yet join them with spaces into a single string? Otherwise, the subsequent shell_escaped_CMDly function (if you really want to split this across two functions) will have a hard time to decide which space in its input is inside a string and which is an argument separator, as it will once more have to implement most of the Microsoft C parsing algorithm to count quotes in order to determine the function of each space.

Reply:
I simply disable it, so there are no permutations to test. The `cmd` parser defines a mapping of a single string to single string, so spaces have no meaning to cmd. Thus it’s unnecessary to quote spaces (although it’s valid to just quote every character). Most resources on this seem to get that fact wrong and confuse `cmd` with the utilities that ship with it (and also with special syntax like `for` and `if`), leading to a confusion of inaccurate claims. To quote correctly, one merely needs to identify the steps in the list that are relevant and perform the inverse list of operations in reverse. That list is actually unexpectedly short—it just needs to prefix a couple characters with `^`.

Reply:
This link is also useful 
https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/ 

Read the conclusions at the end:

> Do:

> 1. Always escape all arguments so that they will be decoded properly by CommandLineToArgvW, perhaps using my ArgvQuote function above.

> 2. After step 1, then if and only if the command line produced will be interpreted by cmd, prefix each shell metacharacter (or each character) with a ^ character.

> Do not:
> 

> 1. Simply add quotes around command line argument arguments without any further processing.
> 2. Allow cmd to ever see an unescaped " character.



Effectively `escape_winsomely` (I also hate this name but w/e)  is step 1 of Raymond Chen's "Do" .    Jameson's CMD escaping  function does step 2.  And as Jameson mentioned we could in principle just escape everything (although overkill) . 

I fail to see why his implementation is incorrect or why the general idea here is incorrect . We are not trying to _duplicate_ how CMD.exe parses the command line.  We are, however, trying to send CMD.exe a valid escaped string from Julia.

Reply:
Looking at an earlier PR https://github.com/JuliaLang/julia/pull/11478 it seems like the approach there is incorrect.